File: haskell/Strategies/MostParts.hs
-- | The most parts strategy keeps a pool of possible secret codes around.
-- After each guess the inconsistent codes are removed from the pool.
--
-- From the pool of all codes a subpool of codes is determined. This subpool
-- contains the codes which would lead to a maximized number of partitions.
-- That means that the best guess would lead to a different (disjoint) set of
-- new consistent codes for each possible answer. If possible this subpool
-- is reduced to consistent guesses.
-- 
-- Additionally, the to-be-entered code which is chosen from the subpool is the
-- one that results in the minimal travel distance for the robot.
-- 
-- If there are less than a handful of codes left the one with the shortest
-- distance is chosen.

-- TODO: Explain the special case when there are only some codes left.
--
-- I can't explain why this special case is needed but without it the strategy
-- is really bad.

-- TODO: Explain why the subpool must be reduced to consistent guesses if
-- possible. Explain why it is not necessarily needed for the other strategies.
--
-- If I don't do this the strategy is bad.

module Strategies.MostParts (mostparts) where

import Data.List (minimumBy, maximumBy, sortBy, groupBy)
import Data.Ord (comparing)
import Data.Function (on)
import Control.Arrow

import Strategies.Util

mostparts = Strategy 
  { initialize   = initializeStrategyState 
  , extractGuess = extractGuess'
  , updateState  = updateState'
  }

extractGuess' :: StrategyState -> Code
extractGuess' = guess

updateState' :: Answer -> StrategyState -> StrategyState
updateState' answer st 
    | isCorrect answer (codeLength st) = reset st 
    | otherwise =
        let lastPos     = getPos $ last $ guess st
            consistent' = getConsistentCodes answer (guess st) (consistent st)
            subpool     = getSubPool (allAnswers st) (allCodes st) consistent'
            subpool'    = getConsistentCodes answer (guess st) subpool
            chosen
                | length consistent' <= codeLength st ^ 2      = consistent'
                | length consistent' == length (consistent st) = consistent'
                | not (null subpool')                          = subpool'
                | otherwise                                    = subpool
            shortest    = getShortestCode lastPos chosen 
        in st { guess = shortest, consistent = consistent' }

getSubPool :: Answers -> Codes -> Codes -> Codes
getSubPool allAnswers allCodes consistent =
  let pairs    = [[ (g, getConsistentCodes a g consistent) 
                    | a <- allAnswers ] 
                    | g <- allCodes   ]
      grouped  = map (groupBy ((==) `on` snd)) pairs        
      parts    = map (getGuess &&& length) grouped          
      sorted   = reverse $ sortBy (comparing snd) parts
      getGuess = fst . head . head
  in map fst $ head $ groupBy ((==) `on` snd) sorted

File: lua/strategies/mostparts.lua
-- A strategy that seeks to maximize the parts of a partition.

local util = require 'strategies.util'
local createAllAnswers = util.createAllAnswers
local createAllCodes = util.createAllCodes
local getConsistentCodes = util.getConsistentCodes
local getShortestCode = util.getShortestCode
local pow = math.pow

-- Associate a number with each button.
local buttonValue = {
  [T[0]] = 0,
  [T[1]] = 1,
  [T[2]] = 2,
  [T[3]] = 3,
  [T[4]] = 4,
  [T[5]] = 5,
  [T[6]] = 6,
  [T[7]] = 7,
}

--- Represent a guess with a unique number.
-- E.g. if guess = (T[0],T[2],T[5],T[7]) return 7520.
local function hashGuess(guess)
  local hash = 0
  for i,button in ipairs(guess) do
    hash = hash + pow(10, i) * buttonValue[button]
  end
  return hash
end

-- Main logic

local allCodes = createAllCodes(CODE_LENGTH)
local allAnswers = createAllAnswers(CODE_LENGTH)
local consistentCodes = {}
local lastGuess = nil
local lastButton = {x=0,y=0}

local function reset()
  for i,code in ipairs(allCodes) do
    consistentCodes[i] = code
  end
  lastGuess = getShortestCode(lastButton, consistentCodes)
  lastButton = lastGuess[CODE_LENGTH]
  return lastGuess
end

local function guess(blacks, whites)
  consistentCodes = getConsistentCodes(consistentCodes, lastGuess, 
      blacks, whites)
  -- This part is essential for this strategy
  if #consistentCodes == 1 then 
    lastGuess = consistentCodes[1]
    lastButton = lastGuess[CODE_LENGTH]
    return lastGuess
  end
  local bestGuesses = {consistentCodes[1]}
  local mostParts = 0
  for _,code in ipairs(allCodes) do
    local parts = 0
    local partSizes = {}
    local partHashes = {}
    local new = true
    for i,answer in ipairs(allAnswers) do
      local partition = getConsistentCodes(consistentCodes, code,
          answer.blacks, answer.whites)
      local size = #partition
      local hash = size > 0 and hashGuess(partition[1]) or 0
      -- A partition is new if its cardinality was not encountered before
      -- and the hash of its first element was not encountered either. This 
      -- distinction is sufficient to distinguish partitions because the 
      -- getConsistentCodes function does not alter the order of the elements 
      -- in the list it returns.
      new = true
      for i=1,#partSizes do
        if size == partSizes[i] and hash == partHashes[i] then
          new = false
          break
        end
      end
      if new then
        partHashes[#partHashes+1] = hash
        partSizes[#partSizes+1] = size
        parts = parts + 1
      end        
    end
    if parts == mostParts and parts > 1 then
      bestGuesses[#bestGuesses+1] = code 
    end
    if parts > mostParts then
      mostParts = parts
      bestGuesses = {code}
    end
  end
  -- Use, if possible, consistent codes
  local consistentBestGuesses = getConsistentCodes(bestGuesses, lastGuess, 
      blacks, whites)
  if #consistentBestGuesses ~= 0 then bestGuesses = consistentBestGuesses end
  -- Use a code with the shortest travel distance
  lastGuess = getShortestCode(lastButton, bestGuesses)
  lastButton = lastGuess[CODE_LENGTH]
  return lastGuess
end

return {
  reset = reset,
  guess = guess
}
